### ISSUE TITLE
Allow abilities per scenario

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#7706]
This was an idea i had while looking at the  the abilities code and the wml code that uses it, and i thought would be rather easy to implement, so i make an issue for possible  thoughts and input:

Allow an [abilities] in [scenario] that would then apply to all units on the map, the abilities can then of course use filters to only apply to certain units/ under certain conditions. 
I think this would make a few things quite simple and direct (for umc developers), that would otherwise requite temporary objects (which by itself isn't that complicated either but the complexity grows as soon as you want to do stuff like for example changing these abilities in the middle of the scenario, deal with umc mods that add/remove units etc.)
Random  usecases i can think of right now:
1) A scenario with locations where all units are invisible, or where all units heal (with more flexibility than villages)
2) Terrains that give weapon  specials like firstrike/berserk etc.
3) A leadership-like ability that works two tiles away.

**Comment by knyghtmare** [comment]
This would be a very welcomed feature. In one of my campaign works, I had plans for an ability that gave all player side 1 units drakes race only, +25% melee dmg + 3 rounds of berserker. Maybe with this feature, it can be implemented more easily then the usual "unit placed" event with `first_time_only=no`.

**Comment by ProditorMagnus** [comment]
Not going to think of specific use cases, but I expect it would help in modification/era as well.

**Comment by knyghtmare** [comment]
I think it would be better to introduce this as `[scenario_unit_effects]` rather than abilities, as using `[abilities]` might cause confusion later.

**Comment by CelticMinstrel** [comment]
I think this could be implemented in Lua using `wesnoth.persistent_tags`. The only part that wouldn't cover is being able to modify the abilities in the middle of a scenario.

**Comment by Pentarctagon** [comment]
Is this different than just using a unit placed event to give units an ability?

**Comment by CelticMinstrel** [comment]
I think it would basically be a shortcut for something similar to that.

**Comment by Pentarctagon** [comment]
Is it really all that useful, in that case? It doesn't seem like adding this feature would save much time/effort over using a unit placed event.

**Comment by CelticMinstrel** [comment]
I'm not sure. I think gfgtdf might have a slightly different idea of what this request means.

**Comment by gfgtdf** [comment]
Well, this doesn't add anything that can not in theory be  already be done in some other way. But in comparision to adding abilities in unit_placed events, its more robust and direct, in particular:
1) you don't have to think about how to remove the ability from all units (for example using object duration might not work on unit that are moved to wml varialbes, also using object duration will rebuild the unit which some umc devs might not want.)
2) It will be easier to change/remove it during a scenario.
3) It is less code for umc devs (and thus easier to use)

> I think it would be better to introduce this as [scenario_unit_effects] rather

I'm not really convinced, abilities already support numerous filters adding another way to wrap them into filter might be more confusing than helpful.

> I think this could be implemented in Lua using wesnoth.persistent_tags.

I would implement this in the c++ side, i am already intending to work on a related part f the code, that's why i brought  this issue up.

**Comment by Pentarctagon** [comment]
I agree that calling it abilities will be confusing. We already have actual abilities and "weapon specials as abilities", so now adding "scenario abilities" is just going to make it less clear what an ability actually is.
